sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java

```diff
@@ -429,6 +434,85 @@ private HttpPipeline setupPipeline(HttpClient httpClient) {
         }).map(ar -> new MsalToken(ar, options));
     }
 
+    /**
+     * Asynchronously acquire a token from Active Directory with a device code challenge. Active Directory will provide
+     * a device code for login and the user must meet the challenge by authenticating in a browser on the current or a
+     * different device.
+     *
+     * @param request the details of the token request
+     * @return a Publisher that emits an AccessToken when the device challenge is met, or an exception if the device
+     *     code expires
+     */
+    public Mono<MsalToken> authenticateWithVsCodeCredential(TokenRequestContext request) {
+
+        VisualStudioCacheAccessor accessor = new VisualStudioCacheAccessor();
+
+        JsonNode userSettings = accessor.getUserSettings();
+
+        String tenant = tenantId;
+        String cloud = "Azure";
+
+        if (!userSettings.isNull()) {
+            if (userSettings.has("azure.tenant")) {
+                tenant = userSettings.get("azure.tenant").asText();
+            }
+
+            if (userSettings.has("azure.cloud")) {
+                cloud = userSettings.get("azure.cloud").asText();
+            }
+        }
+
+        String credential = accessor.getCredentials("VS Code Azure", cloud);
+
+        String authority = KnownAuthorityHosts.AZURE_CLOUD.replaceAll("/+$", "")
+            + "/organizations/" + tenant;
+        PublicClientApplication.Builder publicClientApplicationBuilder = PublicClientApplication.builder(clientId);
+        if (httpPipelineAdapter != null) {
+            publicClientApplicationBuilder.httpClient(httpPipelineAdapter);
+        } else if (options.getProxyOptions() != null) {
+            publicClientApplicationBuilder.proxy(proxyOptionsToJavaNetProxy(options.getProxyOptions()));
+        }
+
+        try {
+            ObjectMapper mapper = new ObjectMapper();
+            JsonNode parsedCredentials = mapper.readTree(credential);
+            String redirectUri = parsedCredentials.get("redirectionUrl").asText();
+            String authorizationCode = parsedCredentials.get("code").asText();
+
+            AuthorizationCodeParameters authorizationCodeParameters = AuthorizationCodeParameters
+                .builder(authorizationCode, new URI(redirectUri))
+                .scopes(new HashSet<>(request.getScopes()))
+                .build();
+
+            PublicClientApplication application = publicClientApplicationBuilder.authority(authority).build();
```

Comment:
In .NET we don't handle the case where the authorization code is stored, only the case where the stored value is a refresh token. We should align this.

Reply:
removed